lib/index.js

```diff
@@ -47,218 +45,201 @@ const loginWeb = (opener, opts = {}) => {
+    const content = await res.json()
     log.verbose('web auth', 'got response', content)
+
+    const { doneUrl, loginUrl } = content
     if (!isValidUrl(doneUrl) || !isValidUrl(loginUrl)) {
       throw new WebLoginInvalidResponse('POST', res, content)
     }
-    return content
-  }).then(({ doneUrl, loginUrl }) => {
-    log.verbose('web auth', 'opening url pair')
-
-    const openPromise = opener(loginUrl, doneEmitter)
-    const webAuthCheckPromise = webAuthCheckLogin(doneUrl, { ...opts, cache: false })
-      .then(authResult => {
-        log.verbose('web auth', 'done-check finished')
-
-        // cancel open prompt if it's present
-        doneEmitter.emit('abort')
-
-        return authResult
-      })
-
-    return Promise.all([openPromise, webAuthCheckPromise]).then(
-      // pick the auth result and pass it along
-      ([, authResult]) => authResult
-    )
-  }).catch(er => {
-    // cancel open prompt if it's present
-    doneEmitter.emit('abort')
 
+    return await webAuthOpener(opener, loginUrl, doneUrl, opts)
+  } catch (er) {
     if ((er.statusCode >= 400 && er.statusCode <= 499) || er.statusCode === 500) {
       throw new WebLoginNotSupported('POST', {
         status: er.statusCode,
         headers: { raw: () => er.headers },
       }, er.body)
-    } else {
-      throw er
     }
-  })
+    throw er
+  }
+}
+
+const webAuthOpener = async (opener, loginUrl, doneUrl, opts) => {
+  const abortController = new AbortController()
+  try {
+    log.verbose('web auth', 'opening url pair')
+    const [, authResult] = await Promise.all([
+      opener(loginUrl, { signal: abortController.signal }),
+      webAuthCheckLogin(doneUrl, { ...opts, cache: false }).then((r) => {
+        log.verbose('web auth', 'done-check finished')
+        abortController.abort()
+        return r
+      }),
+    ])
+    return authResult
+  } catch (er) {
+    abortController.abort()
+    throw er
+  }
 }
 
-const webAuthCheckLogin = (doneUrl, opts) => {
-  return fetch(doneUrl, opts).then(res => {
-    return Promise.all([res, res.json()])
-  }).then(([res, content]) => {
-    if (res.status === 200) {
-      if (!content.token) {
-        throw new WebLoginInvalidResponse('GET', res, content)
-      } else {
-        return content
-      }
-    } else if (res.status === 202) {
-      const retry = +res.headers.get('retry-after') * 1000
-      if (retry > 0) {
-        return sleep(retry).then(() => webAuthCheckLogin(doneUrl, opts))
-      } else {
-        return webAuthCheckLogin(doneUrl, opts)
-      }
-    } else {
+const webAuthCheckLogin = async (doneUrl, opts) => {
+  const res = await fetch(doneUrl, opts)
+  const content = await res.json()
+
+  if (res.status === 200) {
+    if (!content.token) {
       throw new WebLoginInvalidResponse('GET', res, content)
     }
+    return content
+  }
+
+  if (res.status === 202) {
+    const retry = +res.headers.get('retry-after') * 1000
+    if (retry > 0) {
+      await timers.setTimeout(retry)
```

Comment:
One thing I want to confirm is if this will leave the process open if the opener was cancelled. The CLI would always forcefully exit, but anything that would potentially leave the process open is footgun.

Reply:
Reworked this a bit so that the `setTimeout` sets `ref: false` and also gets the signal passed to it in case the opener throws an unrecoverable error. These are all very much edge cases, but just possible enough that I felt better testing them rather than risking them surface via CLI bugs later.